Changes to allow searching and index comparisons.

dmopalmer opened the pull request:
Address issue #2 by adding search and index comparison.

dmopalmer committed:
Changes to allow searching and index comparisons.

zec commented on src/lib.rs:
```diff
@@ -509,6 +511,107 @@ impl<'a, T: CBufItem, const SIZE: usize> CBufReader<'a, T, SIZE> {
+    /// assuming buffer contents are sorted by `predicate()` in `false`, `true` order.
+    ///
+    /// If `fromstart` is set, searches all available data, otherwise
+    /// the search starts at the `current_next_index()`.
```
Nice touch.

zec commented on src/lib.rs:
```diff
@@ -509,6 +511,107 @@ impl<'a, T: CBufItem, const SIZE: usize> CBufReader<'a, T, SIZE> {
+    /// On completion, the `current_next_index()` is set to
+    /// the returned index, or to the end index if
+    /// no true value is found and InvalidIndexError.
+    /// if buffer contents are sorted by `predicate()` in `false`, `true` order.
```
I previously mentioned to you (in a different venue) that I wasn't sure `search` should mutate the `CBufReader`. Looking at this again, I feel that it probably should _not_.

Some explanation is probably in order. I see `CBufReader` as being something you use to read sequentially from the `CBuf`, and this sort of violates the spirit of that by causing the same `CBufReader` to possibly jump back and forth in the buffer.

That said, that leaves the question of what `search` _should_ return instead. How about this idea: on success, it returns a `CBufReader` with `next_local` set to the index `search` currently sets in `self`. Seem reasonable?

dmopalmer replied:
I agree that search should not mutate the `CBufReader`.  Reasonable things it could return on success are:
1. A `CBufIndex`
2. A `CBufReader` copied from `self` and then set with `next_local` at the location.  (Your suggestion.)

I think that I agree that 2. is probably better.

As a nicety, is it possible (good?) for the lifetime of the returned `CBufReader` to not be the lifetime of `self`, but rather be the lifetime of the `CBuf` that `self` got its lifetime from?

zec replied:
@dmopalmer has gone and implemented option (1). That's fine by me.

zec commented on src/lib.rs:
```diff
@@ -509,6 +511,107 @@ impl<'a, T: CBufItem, const SIZE: usize> CBufReader<'a, T, SIZE> {
+    /// (i.e., [`fetch_next_item`](Self::fetch_next_item)
+    /// and [`available_items_iter`](Self::available_items_iter)).
+    #[inline]
+    pub fn set_next_index(&mut self, next: CBufIndex<SIZE>) {
```
I'm not sure whether we should have this method or not. See my comment at [line 532] for more details.

[line 532]: https://github.com/BlackCAT-CubeSat/bc-buf/pull/3/commits/fb6cd628e6749ce98967774d339cd30158d9d29f#diff-b1a35a68f14e696205874893c07fd24fdb88882b47c23cc0e0c80a30c7d53759R532

dmopalmer replied:
Would it be better to replace it with:
```
pub fn reader_starting_at(&self, start: CBufIndex<SIZE>) -> Self;
```

With a lifetime of the returned `CBufReader` that depends on that of the underlying `CBuf`, and not on the `self`.

zec replied:
Yes, and as we discussed yesterday, `Self` is the appropriate return type here.

dmopalmer replied:
Look at my implementation of `reader_starting_at`.  Is the `self.copy()` the right thing?

zec replied:
Per the [standard library docs](https://doc.rust-lang.org/std/clone/trait.Clone.html#derivable), the `Clone` impl `#[derive(Clone)]` generates does a field-wise `Clone`, which is OK in this case.

The bigger issue is that the traits a type implements are part of its public API/ABI. As such, my question is: should `CBufReader` be `Clone` for the crate's users? Answering my own question, given what it is and what it does, I think that's OK, so you're good adding `#[derive(Clone)]` on `CBufReader`.

Incidentally, if you didn't want to have `CBufReader` implement `Clone`, something like the following [would work](https://doc.rust-lang.org/book/ch05-01-defining-structs.html#creating-instances-from-other-instances-with-struct-update-syntax):

```rust
let result = CBufReader {
    next_local: next,
    ..self
};
```

zec reviewed (commented):
I've looked over the changes, and have put comments at the relevant parts of the code.

dmopalmer commented on src/utils.rs:
```diff
@@ -43,6 +43,11 @@ use std::sync::mpsc;
 //   * the next item will be written to cb.buf[idx & (SIZE-1)]
 //   * once the next item has been written, the next value of cb.next will be max(idx.wrapping_add(1), SIZE)
 
+// Ordering and arithmetic
```
I don't understand why having them be ordereable is a bad thing that must be prevented.  Why can't they be ordered?  (Subject to the restriction that it is assumed that they are not more than `usize::MAX/2` apart.).

Being able to tell if one datum comes before another, or before the earliest available data, is very useful in many many use cases.

zec replied:
This disagreement has been resolved. While the `PartialOrd` impl (as it stands now) _does_ necessarily violate [transitivity of &lt;](https://doc.rust-lang.org/core/cmp/trait.PartialOrd.html), it only does so in cases we don't care about (and where [`PartialOrd::partial_cmp`](https://doc.rust-lang.org/core/cmp/trait.PartialOrd.html#tymethod.partial_cmp) returns `None`).

So (with a limit on the distance between indices before they're considered non-ordered with respect to each other) I'm OK with this now.

dmopalmer commented on src/utils.rs:
```diff
@@ -192,6 +205,29 @@ impl<const SIZE: usize> CBufIndex<SIZE> {
+
+    /// Signed value that must be added to `self` to give the other.
+    #[inline]
+    pub fn signed_offset_to(self, other: &Self) -> isize {
```
`Range<idx>` requires that idx is PartialOrd.

https://doc.rust-lang.org/std/ops/struct.Range.html

zec replied:
While I've come around on `PartialOrd`, I'm still uneasy about `signed_offset_to`. Let me think this through some more...

zec replied:
You mentioned yesterday that you'd be OK with removing `signed_offset_to`, and I've done so.

dmopalmer commented on src/lib.rs:
```diff
@@ -509,6 +511,107 @@ impl<'a, T: CBufItem, const SIZE: usize> CBufReader<'a, T, SIZE> {
+        }
+        // At this point, ifalse < itrue and index to false and true values
+
+        while ifalse + 1 < itrue {
```
Without knowing why you don't want `CBufIndex` to be orderable, I don't know know why `bisect` would be better.  

It looks like it might be hard to handle the error condition where the write pointer overtakes the binary search pointer, but calling that case an `InvalidIndexError` seems reasonable.

If the `CBufReader`'s `next_local` is sufficiently stale and `fromstart` is `false` the search will fail immediately if `range` starts with too much overwritten data.  I would begin the range at `next_local.max(self.current_valid_index_range().start)`, but once again, that requires `CBufIndex` to be orderable.


(Looking at the `Range` documentation, it doesn't appear to have the ability to do the intersection of two `Range`s which would a way of disguising the fact that it is using the ordering of the indices.)

zec replied:
Between discussions with you and thinking through it, I'm OK with the comparison operators here.

zec replied:
Apart from removing the commented-out code at [line 576], `search` is good.

[line 576]: https://github.com/BlackCAT-CubeSat/bc-buf/pull/3/files#diff-b1a35a68f14e696205874893c07fd24fdb88882b47c23cc0e0c80a30c7d53759R576

dmopalmer committed:
Search returns an index without mutating the reader.
CBufReader.rewind() added.

zec commented on src/utils.rs:
```diff
@@ -82,6 +87,14 @@ impl<const SIZE: usize> Default for CBufIndex<SIZE> {
     }
 }
 
+/// Ordering for `CBufIndex<SIZE>`
```
I've come around to thinking `PartialOrd` is OK. This implementation isn't quite correct, but I'll fix that myself.

zec replied:
Fixed.

zec committed:
Rewrote `PartialOrd` impl for `CBufIndex`

The "local ordering" of CBufIndex has some subtleties that the previous
impl bulldozed over. This fixes that, and adds tests for assurance.

I also took the opportunity to update the copyright dates and run
rustfmt.

zec committed:
Removed `CBufIndex::signed_offset_to`

This isn't used anywhere in our code, and I'm a little uneasy about
what this might imply about the semantics of `CBufIndex`es, so I'm
removing it.

zec commented on src/lib.rs:
```diff
@@ -529,6 +619,34 @@ impl<'a, T: CBufItem, const SIZE: usize> CBufReader<'a, T, SIZE> {
+    pub fn fetch_with_index(&self, idx: CBufIndex<SIZE>) -> FetchWithIndexResult<T, SIZE> {
+        let mut idx_try = idx;
+        for _ in 0..Self::NUM_READ_TRIES {
+            match self.fetch_seq(idx_try, &()) {
```
Minor nitpick, but it would be better to use `CBufReader::fetch` rather than `CBufReader::fetch_seq` here.

dmopalmer committed:
Fixing up pull request in consultation with Zachary

dmopalmer committed:
CBufReader.search now returns a CBufReader.
Minor additional changes.

zec committed:
Copyediting of doc comments

zec commented on src/lib.rs:
```diff
@@ -509,6 +521,86 @@ impl<'a, T: CBufItem, const SIZE: usize> CBufReader<'a, T, SIZE> {
         self.next_local
     }
 
+    /// New `CBufReader` with the sequential read index at value.
```
Between commits f06816b and 73f36f4, this is resolved.

zec reviewed (approved):
With a couple of iterations of work, this looks good!
